### PR TITLE
Fix pickling for Tensor subclasses (redo)

### LINKED ISSUE #47051: Tensor subclasses lose type when pickling
## 🐛 Bug

Pickling and then unpickling a subclass of `Tensor` should result in an object of the same class as what was pickled. However, it doesn't - instead it always creates something of class `Tensor`.

## To Reproduce

The following prints `torch.Tensor`, however it should print `_T`.

```
class _T(torch.Tensor): ...
t = tensor([1]).as_subclass(_T)
print(type(pickle.loads(pickle.dumps(t))))
```

## Expected behavior

It should print `_T`.

## Environment

`torch.__version__` == 1.7.0, running on Ubuntu.

## Additional context

Placing the following method in the base class resolves the problem:

```
    def __reduce_ex__(self,proto):
        torch.utils.hooks.warn_if_has_hooks(self)
        args = (type(self), self.storage(), self.storage_offset(), tuple(self.size()), self.stride())
        if self.is_quantized: args = args + (self.q_scale(), self.q_zero_point())
        f = _fa_rebuild_qtensor if self.is_quantized else  _fa_rebuild_tensor
        return (f, args + (self.requires_grad, OrderedDict()))
```

Perhaps something like this could be used to fix this issue.

cc @hameerabbasi @rgommers @mruberry

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #47051 
Redo of #47115

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Pending investigation why we broke things.

@hameerabbasi a start to make this patch better, would be to make sure our Pickle format doesn't change for non-subclasses of tensors. This means that when `type is Tensor`, we must not use `_rebuild_from_type` as the rebuild function, as these pickles won't be accessible from earlier version. Also would be great to add big warning blinky lights to this function warning against edits.

**Comment by hameerabbasi** [comment]
Only failed test is on the C++ layer, which wasn't touched in this PR.

File: torch/tensor.py [hunk 1]
```diff
@@ -81,10 +89,16 @@ def __deepcopy__(self, memo):
             return new_tensor
 
     def __reduce_ex__(self, proto):
+        if type(self) is Tensor:
+            return self._reduce_ex_internal(proto)
```
Comment (ezyang) [inline torch/tensor.py:93]: Good

**Comment by hameerabbasi** [comment]
Anything else needed here?

**Comment by hameerabbasi** [comment]
@ezyang This PR could use some attention too.

**Comment by ezyang** [comment]
This one is going to take a little more time because I need to root cause why internal builds failed when we put this change in.

**Review by ezyang — APPROVED** [review]
crypten test no longer fails with the BC-compatible pickle change

**Pull request closed without merging**